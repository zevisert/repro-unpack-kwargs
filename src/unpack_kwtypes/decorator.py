from __future__ import annotations
from abc import abstractmethod

from functools import wraps
from typing import (
    Callable,
    Generic,
    ParamSpec,
    TypeVar,
    cast,
)

P = ParamSpec("P")
R = TypeVar("R", covariant=True)
K = TypeVar("K", contravariant=False, covariant=False)


class DecoratedCallable(Generic[P, R, K]):
    kwtypes: K

    @abstractmethod
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R:
        raise NotImplementedError()


def attach_kwtypes(
    kwtypes: K,
) -> Callable[[Callable[P, R]], DecoratedCallable[P, R, K]]:
    def decorator(func: Callable[P, R]) -> DecoratedCallable[P, R, K]:
        @wraps(func)
        def wrapped(*args: P.args, **kwargs: P.kwargs) -> R:
            return func(*args, **kwargs)

        decorated = cast(DecoratedCallable[P, R, K], wrapped)
        decorated.kwtypes = kwtypes
        return decorated

    return decorator
