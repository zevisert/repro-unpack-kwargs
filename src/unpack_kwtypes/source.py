from itertools import islice
from typing import Literal, LiteralString, TypedDict, Unpack

from unpack_kwtypes.decorator import attach_kwtypes


class Movie(TypedDict):
    title: str
    year: int


MOVIES: list[Movie] = [
    {"title": "The Matrix", "year": 1999},
    {"title": "Avatar", "year": 2009},
    {"title": "Star Wars: Episode IV - A New Hope", "year": 1977},
    {"title": "The Lord of the Rings: The Fellowship of the Ring", "year": 2001},
    {"title": "The Dark Knight", "year": 2008},
    {"title": "The Hunger Games", "year": 2012},
    {"title": "Django Unchained", "year": 2012},
    {"title": "The Godfather", "year": 1972},
    {"title": "The Shawshank Redemption", "year": 1994},
    {"title": "The Wolf of Wall Street", "year": 2013},
    {"title": "The Martian", "year": 2015},
    # Thanks copilot!
]


class QueryKeywordArgs(TypedDict, total=False):
    """Keyword arguments for query_db, to be attached to the function as a kwtypes attribute."""

    limit: int | None
    offset: int | None
    order: Literal["ASC", "DESC"]


@attach_kwtypes(QueryKeywordArgs)
async def query_db(
    title_query: LiteralString,
    /,
    **kwargs: Unpack[QueryKeywordArgs],
) -> list[Movie]:
    matched = sorted(
        [
            movie
            for movie in MOVIES
            if title_query.casefold() in movie["title"].casefold()
        ],
        key=lambda movie: movie["title"],
    )

    if kwargs.get("order", "ASC") == "DESC":
        matched = list(reversed(matched))

    if (offset := kwargs.get("offset", None)) is not None:
        matched = list(islice(matched, offset, None))

    if (limit := kwargs.get("limit", None)) is not None:
        matched = list(islice(matched, limit))

    return matched
