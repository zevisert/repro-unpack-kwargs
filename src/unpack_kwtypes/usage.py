from typing import TYPE_CHECKING, Unpack
from unpack_kwtypes.source import query_db, Movie


async def movie_titles_containing_word_the(
    **kwargs: Unpack[query_db.kwtypes],
) -> list[Movie]:
    return await query_db("the", **kwargs)


if TYPE_CHECKING:
    # src/unpack_kwtypes/usage.py:14: note: Revealed type is "def (**kwargs: Any) -> typing.Coroutine[Any, Any, builtins.list[TypedDict('unpack_kwtypes.source.Movie', {'title': builtins.str, 'year': builtins.int})]]"
    # Expecting: "def (**kwargs: Unpack[TypedDict('unpack_kwtypes.source.QueryKeywordArgs', {'limit'?: Union[builtins.int, None], 'offset'?: Union[builtins.int, None], 'order'?: Union[Literal['ASC'], Literal['DESC']]})]) -> typing.Coroutine[Any, Any, builtins.list[TypedDict('unpack_kwtypes.source.Movie', {'title': builtins.str, 'year': builtins.int})]]"
    reveal_type(movie_titles_containing_word_the)
    # src/unpack_kwtypes/usage.py:17: note: Revealed type is "def (*, limit: Union[builtins.int, None], offset: Union[builtins.int, None], order: Union[Literal['ASC'], Literal['DESC']]) -> TypedDict('unpack_kwtypes.source.QueryKeywordArgs', {'limit'?: Union[builtins.int, None], 'offset'?: Union[builtins.int, None], 'order'?: Union[Literal['ASC'], Literal['DESC']]})"
    # Expecting: "TypedDict('unpack_kwtypes.source.QueryKeywordArgs', {'limit'?: Union[builtins.int, None], 'offset'?: Union[builtins.int, None], 'order'?: Union[Literal['ASC'], Literal['DESC']]})"
    reveal_type(query_db.kwtypes)
