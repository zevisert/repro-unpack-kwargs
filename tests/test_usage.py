import pytest
from syrupy.assertion import SnapshotAssertion

from unpack_kwtypes.usage import movie_titles_containing_word_the


@pytest.mark.asyncio
async def test_nokwargs(snapshot: SnapshotAssertion) -> None:
    await movie_titles_containing_word_the(limit=5) == snapshot


@pytest.mark.asyncio
async def test_limit_order(snapshot: SnapshotAssertion) -> None:
    await movie_titles_containing_word_the(limit=5, order="DESC") == snapshot


@pytest.mark.asyncio
async def test_limit_offset(snapshot: SnapshotAssertion) -> None:
    await movie_titles_containing_word_the(limit=5, offset=5) == snapshot


@pytest.mark.asyncio
async def test_limit_offset_order(snapshot: SnapshotAssertion) -> None:
    await movie_titles_containing_word_the(limit=5, offset=5, order="DESC") == snapshot
