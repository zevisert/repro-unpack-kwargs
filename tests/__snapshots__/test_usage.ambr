# serializer version: 1
# name: test_limit_offset
  list([
    dict({
      'title': 'The Matrix',
      'year': 1999,
    }),
    dict({
      'title': 'The Shawshank Redemption',
      'year': 1994,
    }),
    dict({
      'title': 'The Wolf of Wall Street',
      'year': 2013,
    }),
  ])
# ---
# name: test_limit_offset_order
  list([
    dict({
      'title': 'The Hunger Games',
      'year': 2012,
    }),
    dict({
      'title': 'The Godfather',
      'year': 1972,
    }),
    dict({
      'title': 'The Dark Knight',
      'year': 2008,
    }),
  ])
# ---
# name: test_limit_order
  list([
    dict({
      'title': 'The Wolf of Wall Street',
      'year': 2013,
    }),
    dict({
      'title': 'The Shawshank Redemption',
      'year': 1994,
    }),
    dict({
      'title': 'The Matrix',
      'year': 1999,
    }),
    dict({
      'title': 'The Martian',
      'year': 2015,
    }),
    dict({
      'title': 'The Lord of the Rings: The Fellowship of the Ring',
      'year': 2001,
    }),
  ])
# ---
# name: test_nokwargs
  list([
    dict({
      'title': 'The Dark Knight',
      'year': 2008,
    }),
    dict({
      'title': 'The Godfather',
      'year': 1972,
    }),
    dict({
      'title': 'The Hunger Games',
      'year': 2012,
    }),
    dict({
      'title': 'The Lord of the Rings: The Fellowship of the Ring',
      'year': 2001,
    }),
    dict({
      'title': 'The Martian',
      'year': 2015,
    }),
  ])
# ---
